Relax time handling for PoS minting

domob1812 opened the pull request:
This set of changes fixes a potential overflow with the regtest difficulty and PoS (where the difficulty target is reduced further depending on coin age), and also relaxes handling of the node time when minting PoS blocks.  It does not affect the consensus-level validation logic (except indirectly through the `uint256` changes, which should be reviewed carefully).

Specifically, the miner no longer enforces that the node's local (adjusted) time is beyond the best block; instead, the timestamps of the staking tx and minted block will just be pushed forward as needed, which is possible within some limits.  With this change, it is easily and quickly possible to generate hundreds of PoS blocks on regtest as needed.

domob1812 committed:
Check for overflow with PoS target.

When modifying the PoS target for the coinage, check if the multiply
overflows (and if it does, then it means the target will always be
hit anyway).

This is necessary for proper operation on regtest with minimum difficulty.

domob1812 committed:
Relax time checks when minting PoS blocks.

This adjusts the logic related to the block time when minting PoS blocks
(just how they are mined, *not* the consensus-level verification!).

In particular, we remove the strict check for node time being beyond
the last block timestamp, and instead adjust the time of the staking
tx and block forward as needed to be beyond the current tip.  With this,
it is possible to mine 100 blocks (or a bit more) in one go without
changing the mock time on regtest.

Previously, there was a sleep right in the coinstake creator (that
would try to bring the times up, but wouldn't actually enforce it
anyway).  The real (non-regtest) staking code in miner.cpp already has
a sleep in the loop when it cannot currently stake, which is good
enough.  The setgenerate regtest code now explicitly does a sleep
itself whenever it failed to generate a regtest block (to avoid going
into a busy loop instead).

With these changes, the PowToPosTransition.py test can easily mine
hundreds of PoS blocks quickly; the logic to generate the staking
coin stashes and actually stake (with adjusting the mock time as
necessary) has been moved into two functions.  They will be exposed
as part of the general regtest framework later, so that other
tests can easily generate hundreds of blocks if needed.

galpHub commented on divi/src/uint256.cpp:
```diff
@@ -71,18 +71,41 @@ base_uint<BITS>& base_uint<BITS>::operator*=(uint32_t b32)
+bool base_uint<BITS>::MultiplyBy(const base_uint& b)
 {
+    bool overflow = false;
+    int highestNonzeroA = 0;
+    int highestNonzeroB = 0;
+
```
If we're going to go through the process of multiplying things out and keeping track of the carry term at the end. Having these does not add much does it?

domob1812 replied:
We need them because the inner loop will only process terms that are not going to overflow anyway (so we might miss some overflows, even the unit test fails without those extra checks).

galpHub commented on divi/src/rpcmining.cpp:
```diff
@@ -178,6 +180,10 @@ Value setgenerate(const Array& params, bool fHelp)
                 {
                     blockHashes.push_back(chainActive.Tip()->GetBlockHash().GetHex());
                 }
+            } else if (fProofOfStake)
+            {
+                LogPrintf("Failed to generate PoS block, sleeping\n");
+                MilliSleep(1000);
```
This will likely trigger the timeout on the regression tests if it fails to produce a block, which I think you intended?

domob1812 replied:
Yes indeed.  In fact, I've just now done a follow-up (which we can include here or only later where it is needed) commit that makes `setgenerate` throw an RPC error instead.  The test still goes through even with that.  (But it will e.g. later allow us to test for expected errors, for things like mismatches in the block version.)

Let me know if you want me to merge that into here right now, or just update it later.

domob1812 replied:
I've added in the follow-up commit now as well to this PR.

galpHub commented on divi/qa/rpc-tests/PowToPosTransition.py:
```diff
@@ -27,19 +77,27 @@ def setup_network(self):
+        self.sync_all()
+
+        print ("Trying to mine PoS blocks now...")
+        set_node_times(self.nodes, 0)
```
Is setting the node times here to zero redundant? Considering that we're about to set them when generating the PoS blocks ?

domob1812 replied:
Ah yes indeed, I thought I got rid of all of them - will take a look and remove this.

galpHub commented on divi/src/PoSTransactionCreator.cpp:
```diff
@@ -180,7 +179,12 @@ bool PoSTransactionCreator::PopulateCoinstakeTransaction(
     static int nLastStakeSetUpdate = 0;
     if(!SelectCoins(allowedStakingAmount,nLastStakeSetUpdate,setStakeCoins)) return false;
-    if (GetAdjustedTime() <= chainActive.Tip()->nTime) MilliSleep(10000);
+
+    nTxNewTime = GetAdjustedTime();
+    if (nTxNewTime <= chainActive.Tip()->nTime)
+        nTxNewTime = chainActive.Tip()->nTime + 1;
+    if (nTxNewTime <= chainActive.Tip()->GetMedianTimePast())
+        nTxNewTime = chainActive.Tip()->GetMedianTimePast() + 1;
```
This seems reasonable if just to avoid sleeping where useful work can be done. I'll test it out manually as well.
I would however suggest an alternative:
`nTxNewTime = std::max(GetAdjustedTime(), chainActive.Tip()->GetMedianTimePast() + 1 + nHashDrift);`
see kernel.cpp for the definition of the nHashDrift constant.

The reason is that searching for the hash proof happens in a window of size `nHashDrift` and searched backwards from the initial value of `nTxNewTime`. This will ensure that any PoS found will actually honor the requirement of monotonic `GetMedianTimePast()` values

domob1812 replied:
Ok sure!  I'm not so familiar with the PoS changes yet, but can of course adjust it.

domob1812 replied:
Updated.  On regtest we keep the previous smaller time, so that more blocks can be generated in one go (closer together in time) before we have to adjust the mock time.

domob1812 committed:
Throw in setgenerate if blocks fail.

If generating a new block fails in setgenerate (regtest mode), then
throw an RPC error rather than trying until it may succeed.  This is
cleaner, prevents deadlocks, and also allows to verify for expected
errors (e.g. with wrong block versions or something like that) in
some of the tests.

galpHub reviewed (approved):
Yep. These look good now

domob1812 commented:
Thanks, squashed the two fixup commits.
